Update sample queries for data sharing. (#1103)
Appboy/braze-docs: _docs/_partners/data_and_infrastructure_agility/data_warehouses/snowflake/sample_queries.md

>  This partner page offers some sample queries of possible use cases to reference when setting up your queries.

{% tabs %}
  {% tab Filter By Time%}
  
  A common query might be to filter events by time.
  

>   You can filter them by the time of occurence. Event tables are clustered by `time` which makes filtering by `time` performant.

Reviewer: ```suggestion
  Event tables are clustered by `time` to make filtering on this column performant.
```

Author: Do you think it is worth pointing out what `time` actually means ? Unless it has been done somewhere else.

```
-- find custom events that occured after 04/15/2019 @ 7:02pm (UTC) i.e. timestamp=1555354920
select *
from users_behaviors_customevent_shared
where time > 1555354920
limit 10;
```

>   Or you can filter them by time at which they were persisted in the snowflake datawarehouse i.e. `sf_created_at`. `sf_created_at` and `time` are not the same but usually close, so this query should have similar performance characteristics

Reviewer: ```suggestion
  Or you can filter events by the time at which they were persisted in the Snowflake data warehouse by using `sf_created_at`. `sf_created_at` and `time` are not the same but are usually close, so this query should have similar performance characteristics
```

```
-- find custom events that arrived in Snowflake after time 04/15/2019 @ 7:02pm (UTC)
select *
from users_behaviors_customevent_shared
where sf_created_at > to_timestamp_ntz('2019-04-15 19:02:00')
limit 10;
```
  {% endtab %}
  {% tab Querying Changelogs%}
  
Campaign names and Canvas names are not present in the events themselves. Instead they are published in a changelog table. 

You can see campaign names for events related to a campaign by joining with the campaign changelog table using a query like
```
select e.id, e.time, ccs.time, ccs.name, ccs.conversion_behaviors[e.conversion_behavior_index]
from USERS_CAMPAIGNS_CONVERSION_SHARED e
left join CHANGELOGS_CAMPAIGN_SHARED ccs

> on ccs.id = e.campaign_id

Author: should this be campaign_id or campaign_api_id ?

Reviewer 2: You wrote this. It's campaign_id.

Author: Just making sure.

and ccs.time < e.time
qualify row_number() over (partition by e.id order by ccs.time desc) = 1;
```
Note:
- We are using Snowflake's [window](https://docs.snowflake.com/en/sql-reference/functions-analytic.html) functions here.
- The left join will include events that may not have a `campaign_id`.

> - If you see events with `campaign_api_id`s but no campaign names then there is a possibility that the campaign was created with a name before data sharing existed as a product.

Reviewer: ```suggestion
- If you see events with `campaign_api_id`s but no campaign names then there is a possibility that the campaign was created with a name before Data Sharing existed as a product.
```

- You can see canvas names using a similar query, joining with the `CHANGELOGS_CANVAS_SHARED` table instead.

If you want to see both campaign and canvas names, you may have to use a sub-query as shown below.
```
select campaign_join.*, canvas.name as canvas_name
from 
(SELECT e.id AS event_id, e.external_user_id, e.time, e.user_id, e.device_id, e.sf_created_at,
    e.campaign_api_id, e.canvas_api_id, e.canvas_step_api_id, 
    campaign.name AS campaign_name
  FROM USERS_MESSAGES_INAPPMESSAGE_CLICK_SHARED AS e
  LEFT JOIN CHANGELOGS_CAMPAIGN_SHARED AS campaign ON campaign.id = e.campaign_id
  WHERE e.time >= 1574830800 AND e.time <= 1575176399
  qualify row_number() over (partition by e.id order by campaign.time desc) = 1) as campaign_join

> left join CHANGELOGS_CANVAS_SHARED AS canvas ON canvas.id = campaign_join.canvas_api_id

Author: just want to be sure if we are joining  on *_api_id or *_id

Reviewer 2: same, this should be canvas_id. You can also see this in Snowflake yourself. Did you not run the queries to verify?

qualify row_number() over (partition by campaign_join.event_id order by canvas.time desc) = 1;
```
  {% endtab %}
  {% tab Push Funnel %}

  You can use this Push Funnel query to aggregate push sends raw event data, through to deliveries raw event data, through to opens raw event data. This query shows how all the tables should be joined, since each raw event typically has a separate table.

```sql

SELECT
    COUNT(DISTINCT s."ID" ) AS "users_messages_pushnotification_send.push_sent",
    COALESCE((COUNT(DISTINCT s."ID" )),0)-COALESCE((COUNT(DISTINCT b."ID" )),0) AS "users_messages_pushnotification_send.push_delivered",
    COUNT(DISTINCT o."ID" ) AS "users_messages_pushnotification_open.push_opens"

> FROM users_messages_pushnotification_send_shared AS s

Reviewer 2: since this is documentation, can we keep it at least somewhat verbose? like push_send and push_open, instead of s and o

Author: send and open seem a good middleground.

>  This partner page offers some sample queries of possible use cases to reference when setting up your queries.

{% tabs %}
  {% tab Filter By Time%}
  
  A common query might be to filter events by time.
  
  You can filter them by the time of occurrence. Event tables are clustered by `time` which makes filtering by `time` performant.

```sql
-- find custom events that occurred after 04/15/2019 @ 7:02pm (UTC) i.e. timestamp=1555354920
SELECT *
FROM users_behaviors_customevent_shared
WHERE time > 1555354920
LIMIT 10;
```

>   Or you can filter events by the time at which they were persisted in the Snowflake data warehouse by using `sf_created_at`. `sf_created_at` and `time` are not the same but are usually close, so this query should have similar performance characteristics

Reviewer 2: You might have to add a note here saying that SF_CREATED_AT is available for events that happened on this date and beyond. And that rest will be NULL.

Author: `sf_created_at` has valid values for events that were persisted in the data warehouse around `2019-09-17` and onwards. Events before that have sf_created_at set to null.

Is that ok ? The exact time at which sf_created_at starts is different for each table.

Reviewer 2: No, the date is Nov 16th and beyond since SF_CREATED_AT was broken for few weeks.

https://jira.braze.com/browse/DI-1196

Author: Updated. Perhaps we should set `sf_created_at` for events before that to null to reduce confusion.

```sql
-- find custom events that arrived in Snowflake after time 04/15/2019 @ 7:02pm (UTC)
SELECT *
FROM users_behaviors_customevent_shared
WHERE sf_created_at > to_timestamp_ntz('2019-04-15 19:02:00')
LIMIT 10;
```
  {% endtab %}
  {% tab Querying Changelogs%}
  
Campaign names and Canvas names are not present in the events themselves. Instead, they are published in a changelog table. 

You can see campaign names for events related to a campaign by joining with the campaign changelog table using a query like

```sql
SELECT e.id, e.time, ccs.time, ccs.name, ccs.conversion_behaviors[e.conversion_behavior_index]
FROM USERS_CAMPAIGNS_CONVERSION_SHARED e
LEFT JOIN CHANGELOGS_CAMPAIGN_SHARED ccs
ON ccs.id = e.campaign_id
AND ccs.time < e.time
qualify row_number() over (partition by e.id ORDER BY ccs.time DESC) = 1;
```
Note:
- We are using Snowflake's [window](https://docs.snowflake.com/en/sql-reference/functions-analytic.html) functions here.

> - The left join will include events that may not have a `campaign_id`.

Reviewer 2: not sure what you mean here?

Author: If there are events in the table not associated with a campaign, and if they want to see them, left join will be useful.

Reviewer 2: ah okay

> - If you see events with `campaign_api_id`s but no campaign names then there is a possibility that the campaign was created with a name before Data Sharing existed as a product.

Reviewer 2: campaign_id

      (SELECT DISTINCT email_address AS click_address,
      message_variation_api_id AS c_message_variation_api_id,
      canvas_step_api_id AS c_canvas_step_api_id
      FROM USERS_MESSAGES_EMAIL_CLICK_SHARED)      SELECT * FROM deliveries
      LEFT JOIN opens
      ON (deliveries.delivered_address)=(opens.open_address)
      AND ((deliveries.d_message_variation_api_id)=(opens.o_message_variation_api_id) OR (deliveries.d_canvas_step_api_id)=(opens.o_canvas_step_api_id))
      LEFT JOIN clicks
      ON (deliveries.delivered_address)=(clicks.click_address)
      AND ((deliveries.d_message_variation_api_id)=(clicks.c_message_variation_api_id) OR (deliveries.d_canvas_step_api_id)=(clicks.c_canvas_step_api_id))
      )
SELECT
    email_messaging_cadence."DIFF_DAYS"  AS "email_messaging_cadence.days_since_last_received",
    (count(distinct email_messaging_cadence."OPEN_ADDRESS", email_messaging_cadence."O_MESSAGE_VARIATION_API_ID")
      +count(distinct email_messaging_cadence."OPEN_ADDRESS", email_messaging_cadence."O_CANVAS_STEP_API_ID"))/(COUNT(DISTINCT email_messaging_cadence."DELIVERED_ID" ))  AS "email_messaging_cadence.unique_open_rate"

> FROM email_messaging_cadenceGROUP BY 1

Reviewer 2: space before GROUP BY
